qiskit/transpiler/passes/layout/sabre_layout.py

```diff
@@ -0,0 +1,148 @@
+# that they have been altered from the originals.
+
+"""Layout selection using the SABRE bidirectional search approach from Li et al.
+"""
+
+import logging
+import numpy as np
+
+from qiskit.transpiler.layout import Layout
+from qiskit.transpiler.basepasses import AnalysisPass
+from qiskit.transpiler.exceptions import TranspilerError
+
+logger = logging.getLogger(__name__)
+
+
+class SabreLayout(AnalysisPass):
+    """Choose a Layout via iterative bidirectional routing of the input circuit.
+
+    Starting with a random initial `Layout`, the algorithm does a full routing
+    of the circuit (via the `routing_pass` method) to end up with a
+    `final_layout`. This final_layout is then used as the initial_layout for
+    routing the reverse circuit. The algorithm iterates a number of times until
+    it finds an initial_layout that reduces full routing cost.
+
+    This method exploits the reversibility of quantum circuits, and tries to
+    include global circuit information in the choice of initial_layout.
+
+    **References:**
+
+    [1] Li, Gushu, Yufei Ding, and Yuan Xie. "Tackling the qubit mapping problem
+    for NISQ-era quantum devices." ASPLOS 2019.
+    `arXiv:1809.02573 <https://arxiv.org/pdf/1809.02573.pdf>`_
+    """
+
+    def __init__(self, coupling_map, routing_pass=None, seed=None,
+                 max_iterations=3):
+        """SabreLayout initializer.
+
+        Args:
+            coupling_map (Coupling): directed graph representing a coupling map.
+            routing_pass (BasePass): the routing pass to use while iterating.
+            seed (int): seed for setting a random first trial layout.
+            max_iterations (int): number of forward-backward iterations.
+        """
+        super().__init__()
+        self.coupling_map = coupling_map
+        self.routing_pass = routing_pass
+        self.seed = seed
+        self.max_iterations = max_iterations
+
+    def run(self, dag):
+        """Run the SabreLayout pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): DAG to find layout for.
+
+        Raises:
+            TranspilerError: if dag wider than self.coupling_map
+        """
+        from qiskit.converters import dag_to_circuit
+        from qiskit.transpiler.passes.layout.set_layout import SetLayout
+        from qiskit.transpiler.passes.layout.full_ancilla_allocation import FullAncillaAllocation
+        from qiskit.transpiler.passes.layout.enlarge_with_ancilla import EnlargeWithAncilla
+        from qiskit.transpiler.passes.layout.apply_layout import ApplyLayout
+        from qiskit.transpiler.passes.routing import SabreSwap
+        from qiskit.transpiler.passmanager import PassManager
+
+        def _layout_and_route_passmanager():
+            """Return a passmanager for a full layout and routing.
+
+            We use a factory to remove potential statefulness of passes.
+            """
+            layout_and_route = [SetLayout(initial_layout),
```

Comment:
considering passing this `initial_layout` as a parameter to the function. Otherwise, the scope gets confusing.

Reply:
done thanks